### PR TITLE
Added a Timesketch enhancer module, pt 1

### PR DESCRIPTION
**Pull request by kiddinn**
This PR adds a new Timesketch enhancer module.

The enhancer module should be run after Timesketch has been run, what it does is:

+ Waits for all analyzers to complete their run
+ Once analyzers are done, collects information from TS, such as views, aggregations and stories
+ Creates a report and stores containers.

The information from the enhancer can then be used later by other modules, such as for SIEM integration and or other form of reporting.

This feature will be split into three PRs, this is part I.

Part 1: The enhancer itself, and utilities
Part 2: Request Tracker PoC exporter and a recipe that uses it
Part 3: Unit tests and other clean-up

### CODE REVIEW

File: dftimewolf/lib/enhancers/timesketch.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+    self._WaitForAnalyzers(sketch)
+
+    # Force a refresh of sketch data.
+    _ = sketch.lazyload_data(refresh_cache=True)
+
+    summary_lines.append(self._formatter.IndentStart())
+
+    view_string = self._GenerateViewString(sketch)
+    if view_string:
+      summary_lines.append(self._formatter.IndentText(
```
Comment (tomchop) [inline dftimewolf/lib/enhancers/timesketch.py:331]: for easier debugging, please store the result of the `_formatter` into a variable and call `append` outside the `if` block. Same for other constructs like this below.

File: dftimewolf/lib/enhancers/timesketch.py [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+              view_string,
+              self._formatter.IndentEnd())))
+    else:
+      summary_lines.append(self._formatter.IndentText(
+          'No views were generated by analyzers.'))
+
+    aggregation_string = self._GenerateAggregationString(sketch)
+    if aggregation_string:
+      summary_lines.append(self._formatter.IndentText(
+          'Following aggregations were discovered:\n{0:s}{1:s}{2:s}'.format(
```
Comment (tomchop) [inline dftimewolf/lib/enhancers/timesketch.py:344]:
```suggestion
          'The following aggregations were discovered:\n{0:s}{1:s}{2:s}'.format(
```

File: dftimewolf/lib/enhancers/timesketch.py [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+            views_to_skip='',
+            aggregations_to_skip='',
+            include_stories=False,
+            token_password='',
+            max_checks=0,
+            formatter='html'):
+    """Sets up a Timesketch Enhancer module.
+
+    Args:
+      timesketch_quick (bool): If set to True then the enhancer will not wait
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:56]: ack, agree. ... changed

File: dftimewolf/lib/enhancers/timesketch.py [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+          until all analyzers are done running (and therefore skip adding
+          reports).
+      views_to_skip (str): A JSON formatted string with a list of View names
+          that are not to be included when generating reports.
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:60]: SG

File: dftimewolf/lib/enhancers/timesketch.py [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+    if aggregations_to_skip:
+      self._aggregations_to_skip = [
+          str(x) for x in json.loads(aggregations_to_skip)]
+    else:
+      self._aggregations_to_skip = []
+
+    if views_to_skip:
+      self._views_to_skip = [str(x) for x in json.loads(views_to_skip)]
+    else:
+      self._views_to_skip = []
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:102]: done

File: dftimewolf/lib/enhancers/timesketch.py [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+
+    if formatter.lower() == 'markdown':
+      self._formatter = utils.MarkdownFormatter()
+    else:
+      self._formatter = utils.HTMLFormatter()
+
+    # Check that we have a timesketch session.
+    if not (self.timesketch_api or self.timesketch_api.session):
+      message = 'Could not connect to Timesketch server'
+      self.ModuleError(message, critical=True)
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:112]: no, this is repetitive  ...

File: dftimewolf/lib/enhancers/timesketch.py [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+
+  def Process(self):
+    """Executes a Timesketch enhancer module."""
+    if self._timesketch_quick:
+      self.logger.info('Not waiting for analyzers to run.')
+      return
+
+    if not self.timesketch_api:
+      message = 'Could not connect to Timesketch server'
+      self.ModuleError(message, critical=True)
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:303]: it shouldn't.... this is an `extra` check to make sure... maybe if another module somewhere does `delete` it by accident, should ont happen...

want me to get rid of this?

File: dftimewolf/lib/enhancers/timesketch.py [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+
+    sketch = self.state.GetFromCache('timesketch_sketch')
+    if not sketch:
+      message = 'Missing the sketch, unable to connect to Timesketch.'
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:307]: could be both, depends on the error message, let me make this more clear.

File: dftimewolf/lib/enhancers/timesketch.py [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+      if status_set.issubset(self._ANALYZERS_COMPLETE_SET):
+        break
+
+      check_number += 1
+      time.sleep(self._ANALYZER_SECONDS_BETWEEN_CHECK)
+
+  def Process(self):
+    """Executes a Timesketch enhancer module."""
+    if self._timesketch_quick:
+      self.logger.info('Not waiting for analyzers to run.')
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:298]: doen

File: dftimewolf/lib/enhancers/timesketch.py (hunk 1 above)
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:331]: ack

File: dftimewolf/lib/enhancers/timesketch.py (hunk 2 above)
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:344]: done

File: dftimewolf/lib/enhancers/timesketch.py [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+          level=2))
+
+      if aggregation.name in self._aggregations_to_skip:
+        continue
+
+      data_frame = aggregation.table
+      nr_rows, _ = data_frame.shape
+      # Check whether a data frame is empty or not.
+      if not nr_rows:
+        continue
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:143]: yes, I have found this to be a more reliable way of determining it, but I think that is no longer the case, don't mind which one I use, if you have strong feelings one way or another

File: dftimewolf/lib/enhancers/timesketch.py [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+      message = 'Could not connect to Timesketch server'
+      self.ModuleError(message, critical=True)
+
+  def _GetSketchURL(self, sketch):
+    """Returns a URL to access a sketch."""
+    api_root = sketch.api.api_root
+    ts_url, _, _ = api_root.partition('/api/v1')
+    return '{0:s}/sketch/{1:d}/'.format(ts_url, sketch.id)
+
+  def _GenerateAggregationString(self, sketch):
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:113]: sure... let's split this up a bit..... SGTM

File: dftimewolf/lib/enhancers/timesketch.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:143]: done

File: dftimewolf/lib/enhancers/timesketch.py [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+      data_frame.drop(['bucket_name'], axis=1, inplace=True)
+      columns = list(data_frame.columns)
+      # We are presenting aggregations here, which is a table that consists of
+      # a column name and then the count. For easier reading in reports we want
+      # the count to be the last column displayed.
+      try:
+        count_index = columns.index('count')
+        count = columns.pop(count_index)
+        columns.append(count)
+      except ValueError:
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:154]: in the case where "count" does not exist... then `columns.index('count')` will fail, and in that case we just want to ignore this whole thing

File: dftimewolf/lib/enhancers/timesketch.py [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+        pass
+
+      self.state.StoreContainer(containers.DataFrame(
+          data_frame=data_frame[columns],
+          description='Timesketch Aggregation: {0:s}'.format(
+              aggregation.name), name=self._REPORT_NAME))
+
+    return '\n'.join(aggregation_strings)
+
+  def _GenerateStoryString(self, sketch):
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:164]: I changed htem all

File: dftimewolf/lib/enhancers/timesketch.py [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+
+      view_url = '{0:s}explore?view={1:d}'.format(sketch_url, view.id)
+      view_strings.append(self._formatter.IndentText(
+          self._formatter.Link(url=view_url, text=view.name), level=2))
+
+      data_frame = sketch.explore(view=view, as_pandas=True)
+
+      nr_rows, _ = data_frame.shape
+      if not nr_rows:
+        continue
```
[the remark this answers is not in the document]
Comment (kiddinn) [reply dftimewolf/lib/enhancers/timesketch.py:235]: do you feel strongly about it?

**Comment by kiddinn** [comment]
@tomchop done with this round, PTAL

File: dftimewolf/lib/enhancers/timesketch.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tomchop) [reply dftimewolf/lib/enhancers/timesketch.py:338]: Yeah, I find it a bit confusing to have this check here if it's not necessary.

File: dftimewolf/lib/enhancers/timesketch.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (tomchop) [reply dftimewolf/lib/enhancers/timesketch.py:160]: If you think `empty()` is as reliable as counting the number of rows, then I'd rather you'd use that instead

File: dftimewolf/lib/enhancers/timesketch.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (tomchop) [reply dftimewolf/lib/enhancers/timesketch.py:278]: I feel a moderate amount of force :P

**Review by tomchop — CHANGES_REQUESTED** [review]

**Comment by kiddinn** [comment]
@tomchop PTAL

**Review by tomchop — APPROVED** [review]

**Pull request merged**